janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -414,63 +423,50 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
      * to how driver paging metadata is stored. The driver stores a full history of a given query's paging metadata
-     * which can lead to OOM issues on non-trivially sized data sets. This class overcomes this by doing another level
-     * of paging that re-executes the query after a configurable number of rows. When the original query is re-executed
-     * it is initialized to the correct offset using the last page's metadata.
+     * which can lead to OOM issues on non-trivially sized data sets. This class overcomes this by forcing the internal
+     * metadata which isn't needed anymore to be deleted to be free for GC.
      */
     private class CQLPagingIterator implements Iterator<Row> {
 
-        private ResultSet currentResultSet;
-
-        private int index;
-        private int paginatedResultSize;
-        private final Supplier<Statement> statementSupplier;
-
-        private byte[] lastPagingState = null;
+        private final List<ExecutionInfo> dseStoredExecutionInfos;
+        private final java.util.Iterator<Row> currentPageIterator;
 
-        public CQLPagingIterator(final int pageSize, Supplier<Statement> statementSupplier) {
-            this.index = 0;
-            this.paginatedResultSize = pageSize;
-            this.statementSupplier = statementSupplier;
-            this.currentResultSet = getResultSet();
+        public CQLPagingIterator(BoundStatement boundStatement) {
+            ResultSet currentResultSet = session.execute(boundStatement);
+            currentPageIterator = currentResultSet.iterator();
+            this.dseStoredExecutionInfos = currentResultSet.getExecutionInfos();
         }
 
         @Override
         public boolean hasNext() {
-            return !currentResultSet.isExhausted();
+            return currentPageIterator.hasNext();
         }
 
         @Override
         public Row next() {
-            if(index == paginatedResultSize) {
-                currentResultSet = getResultSet();
-                this.index = 0;
-            }
-            this.index++;
-            lastPagingState = currentResultSet.getExecutionInfo().getPagingStateUnsafe();
-            return currentResultSet.one();
-
-        }
-
-        private ResultSet getResultSet() {
-            final Statement boundStmnt = statementSupplier.get();
-            if (lastPagingState != null) {
-                boundStmnt.setPagingStateUnsafe(lastPagingState);
+            if(dseStoredExecutionInfos.size()>1){
+                if(dseStoredExecutionInfos.size()==2){
+                    dseStoredExecutionInfos.set(0, dseStoredExecutionInfos.remove(1));
+                } else {
+                    ExecutionInfo lastExecutionInfo = dseStoredExecutionInfos.get(dseStoredExecutionInfos.size()-1);
+                    dseStoredExecutionInfos.clear();
+                    dseStoredExecutionInfos.add(lastExecutionInfo);
+                }
```

Comment:
This part should be unreachable in most cases. The possible case I think about when this part is actually reachable is when we have `storage.page-size=1` and call `next() & hasNext() & next()`. In such case it's possible that during the last `next()` operation our `dseStoredExecutionInfos.size()` will be equal to `3`. In normal scenarios when we always call `hasNext()` before `next()` we shouldn't reach this `else` case.